### PR TITLE
Import rust-installer & adjust compression settings

### PR DESCRIPTION
**Pull request by Mark-Simulacrum**
This brings in rust-lang/rust-installer#123, which enables a larger compression window (8 -> 64MB) amongst other changes to the xz compression settings. The net effect should be smaller compressed tarballs which will decrease bandwidth usage for
static.rust-lang.org, download times, and decompression time.

This comes at the cost of higher baseline requirements for running rustup to use these files, which we believe should be largely acceptable (running rustc is likely to use at least this much memory) but if we get specific reports we may explore options to decrease impact (e.g., using the gzip tarballs automatically in rustup).

To simplify iteration on compression settings this also imports the rust-lang/rust-installer submodule, it is now hosted fully inside rust-lang/rust. Once we land this I'll file a followup to add a note to that repo and we can subsequently archive it.

--

CI times for dist-x86_64-linux builds:

* threads=6, master - 2h 50m
* threads=1, new - 3h 40m
* threads=6, new - 2h 50m

### CODE REVIEW

**Comment by Mark-Simulacrum** [comment]
@bors r-

I want to check locally with dev-static

**Comment by Mark-Simulacrum** [comment]
Baseline install of nightly-2023-02-27 takes 53 MB with `RUSTUP_IO_THREADS=1 RUSTUP_UNPACK_RAM=33554432 rustup update nightly-2023-02-27`. Installing the nightly from the try build with the same environment settings takes 102 MB. So we've increased max rss for a nightly install by around 50 MB. Overall I think this fits within the parameters we were aiming for in infra discussion.

Wall-time for installation seems to take roughly the same amount of time with the minimum settings (6.9s before, 6.4s now). When not constraining I/O threads or memory, 7.2s before, 6.5s now.

Ultimately this all seems fairly reasonable so going to go ahead and move ahead, if we run into trouble we can back it out.

@bors r=pietroalbini

**Comment by matthiaskrgr** [comment]
@bors rollup=never
it looks like this blows up ci times

master:
 [TIMING] dist::PlainSourceTarball PlainSourceTarball -- 123.972
 
 my rollup which included this pr:
 [TIMING] dist::PlainSourceTarball PlainSourceTarball -- 1442.916
 
 this is like more then factor 10 for just this tarball? Just compression takes almost half an hours now.

**Comment by Mark-Simulacrum** [comment]
Hm, probably due to dropping the thread count. I'll update to avoid that.

File: src/bootstrap/test.rs [hunk 1]
```diff
@@ -962,7 +962,7 @@ fn compare_browser_ui_test_version(installed_version: &str, src: &Path) {
         Ok(v) => {
             if v.trim() != installed_version {
                 eprintln!(
-                    "⚠️ Installed version of browser-ui-test (`{}`) is different than the \
+                    "â ï¸ Installed version of browser-ui-test (`{}`) is different than the \
```
Comment (klensy) [inline src/bootstrap/test.rs:965]: Broken

File: src/bootstrap/test.rs [hunk 2]
```diff
@@ -2549,7 +2549,7 @@ impl Step for Bootstrap {
             .env("RUSTC", &builder.initial_rustc);
         if let Some(flags) = option_env!("RUSTFLAGS") {
             // Use the same rustc flags for testing as for "normal" compilation,
-            // so that Cargo doesn’t recompile the entire dependency graph every time:
+            // so that Cargo doesnât recompile the entire dependency graph every time:
```
Comment (klensy) [inline src/bootstrap/test.rs:2552]: and here

**Comment by klensy** [comment]
Feels wrong that this thing not broken
https://github.com/rust-lang/rust/blob/5423745db8b434fcde54888b35f518f00cce00e4/src/bootstrap/tool.rs#L425

**Comment by Mark-Simulacrum** [comment]
External tool controls things like lints, it should be fine to leave that as is. Need to fix the tests though.

**Comment by Mark-Simulacrum** [comment]
@pietroalbini Alright, I think this is good to go now - tests seem to pass and CI times from try build with n=6 threads also match pre-change times.

I'll probably look at some cleanup (rustfmt, lints, maybe inlining the whole code into rustbuild proper) in future PRs but this seems good for now.

File: src/tools/rust-installer/.gitignore [hunk 3]
```diff
@@ -0,0 +1,5 @@
+*~
+tmp
+target/
+**/*.rs.bk
+Cargo.lock
```
Comment (klensy) [inline src/tools/rust-installer/.gitignore:5]: Add `rust-installer` to https://github.com/rust-lang/rust/blob/master/Cargo.toml or keep `Cargo.lock`?

Comment (Mark-Simulacrum) [reply src/tools/rust-installer/.gitignore:5]: It's already a member of the top-level workspace. We could adjust the ignore file here to remove these but I don't think it's worth worrying over.

**Comment by klensy** [comment]
> Feels wrong that this thing not broken
> 
> https://github.com/rust-lang/rust/blob/5423745db8b434fcde54888b35f518f00cce00e4/src/bootstrap/tool.rs#L425
> 
> https://github.com/rust-lang/rust/blob/5423745db8b434fcde54888b35f518f00cce00e4/src/bootstrap/lib.rs#L486

This comment strike back: while working on #112884 i found out that this flag set `SourceType::Submodule` for rust-installer and this is cause why rustfmt didn't format it's sources. It seems there no working assert that submodule is actually submodule.

**Pull request merged**